_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+If not detected and fixed in time, this may lead to disk space exhaustion and service disruption.
+
+Fortunately, there are embedded tools that can be used to find all hanging transactions and rollback them.
+After that, the LSO starts to increment again on every transaction completion and the application should be able to recover.
+
+```sh
+# find all hanging transactions in a broker
+$ kafka-transactions.sh --bootstrap-server :9092 find-hanging --broker 0
+Topic                  Partition   ProducerId  ProducerEpoch   StartOffset LastTimestamp               Duration(s)
+__consumer_offsets     27          171100      1               913095344   2022-06-06T03:16:47Z        209793
+
+# abort an hanging transaction
+$ kafka-transactions.sh --bootstrap-server :9092 abort --topic __consumer_offsets --partition 27 --start-offset 913095344
+```
+
+### Basic example
+
+Let's now run a basic example to see how transactions work at the partition level.
+You just need a computer with few essential command line tools.
+
+This application consumes text messages from the `input-topic`, reverts their content and sends the result to the `output-topic`.
+
+![txn app](/assets/images/posts/2023-04-14-kafka-txn-app.png)
+
+In the following snippet we start the cluster, run the application and create a single transaction.
+For the sake of simplicity, we use a single-node Kafka cluster running on localhost, but the example also works with a multi-node cluster.
+
+```sh
+# download the Apache Kafka binary and start a single-node Kafka cluster in background
+$ KAFKA_VERSION="3.4.0" KAFKA_HOME="$(mktemp -d -t kafka.XXXXXXX)" PATH="$KAFKA_HOME/bin:$PATH" && export KAFKA_HOME PATH
+
+$ curl -sLk "https://archive.apache.org/dist/kafka/$KAFKA_VERSION/kafka_2.13-$KAFKA_VERSION.tgz" \
+  | tar xz -C "$KAFKA_HOME" --strip-components 1
+
+$ zookeeper-server-start.sh -daemon $KAFKA_HOME/config/zookeeper.properties \
+  && sleep 5 && kafka-server-start.sh -daemon $KAFKA_HOME/config/server.properties
+
+# open a new terminal, get and run the application (Ctrl+C to stop)
+$ git clone git@github.com:fvaleri/examples.git -n --depth=1 --filter=tree:0 && cd examples \
+  && git sparse-checkout set --no-cone kafka/kafka-txn && git checkout 5907e0bd31175b5f411dc0e0804cf69cde76ec90
+...
+HEAD is now at 5907e0b Add SerializationException
+
+$ export BOOTSTRAP_SERVERS="localhost:9092" INSTANCE_ID="kafka-txn-0" \
+  GROUP_ID="my-group" INPUT_TOPIC="input-topic" OUTPUT_TOPIC="output-topic"
+
+$ mvn clean compile exec:java -q -f kafka/kafka-txn/pom.xml
+Starting instance with TID kafka-txn-0
+Created topics: input-topic
+Waiting for new data
+...
+
+# come back to the previus terminal and send a test message
+$ kafka-console-producer.sh --bootstrap-server :9092 --topic input-topic
+>test
+>^C
+```
+
+All of this is pretty boring, but now comes the interesting part.
```

Comment:
`boring` -> `normal`/`classic` or even only keep `Now comes the interesting part.`